rfcs/029-work_state_modelling/README.md

```diff
@@ -0,0 +1,177 @@
+
+3) There are already 2 type parameters on `WorkData`, both for the identified state of things such as concepts, and also for the identified state of images. Also, if we decided to add relation data directly to `WorkData` (as optional fields) rather than creating new work types we would need an additional third type parameter (to express whether the referenced works are pre or post minter). These are all tied together so ideally we should be able to specify them with a single type parameter.
+
+4) The type hierarchy from `BaseWork` downwards is quite complex and it can be hard to understand the meaning of particular types of work within the code. The naming could also do with changes in places (e.g. `TransformedBaseWork` does not make much sense as a name, given that the concept of a work does not exist before the transformer).
+
+## Proposed Solution
+
+This RFC proposes having a cleaner separation of the pipeline state of a work with the sort of work it is, namely with different forms of polymorphism:
+
+1) Sub-typing is used for differentiating different sorts of work, with a `Work` being a sum type ADT consisting of the finite number of possible case classes. Note this proposed sub-typing is much simpler than the current implementation, having only a single level of depth and a single parent type.
+
+2) Parameterisation of the `Work` by a state parameter indicates what stage of the pipeline it is in, and which can hold specific data depending on stage. By constraining the parameter to one of a set of known types, we can consider the `Work` model as a finite state machine with the stages in our pipeline each containing a particular transition.
+
+By separating these two things like this we are able to add new types of work or new pipeline states without worrying about how one will impact the other. The following is a sketch of how this might look:
+
+```scala
+/** WorkState represents the state of the work in the pipeline, and contains
+  * different data depending on what state it is. This allows us to consider the
+  * Work model as a finite state machine with the following stages corresponding
+  * to stages of the pipeline:
+  *
+  *      |
+  *      | (transformer)
+  *      ▼
+  *   Unmerged
+  *      |
+  *      | (matcher / merger)
+  *      ▼
+  *   Merged
+  *      |
+  *      | (relation embedder)
+  *      ▼
+  * Denormalised
+  *      |
+  *      | (id minter)
+  *      ▼
+  *  Identified
+  *
+  * Each WorkState also has an associated IdentifierState which indicates whether
+  * the corresponding WorkData is pre or post the minter.
+  */
+sealed trait WorkState {
+  type Id <: IdentifierState
+  type ImageId <: IdentifierState
+}
+
+object WorkState {
+
+  case class Unmerged() extends WorkState {
+    type Id = IdentifierState.Unminted
+    type ImageId = IdentifierState.Unidentified
+  }
```

Comment:
This is how things are currently. `Unminted` means it can be `Unidentified` or `Unidentifiable`. Images are always unidentifiable (unlike concepts etc), so it doesn't make sense to use `Unminted` for that. Naming is slightly weird, I agree, but that's the best we could come up with at the time.

Reply:
@warrd assuming you mean "images are always _identifiable_ here"

Reply:
Yep!